### PR TITLE
Grant read access to additional users on image repository creation,

### PR DESCRIPTION
**Pull request by rcerven**
based on config map in the user namespace named "image-controller-additional-users"

[STONEBLD-2666](https://issues.redhat.com//browse/STONEBLD-2666)

### CODE REVIEW

**Review by mmorhun — COMMENTED** [review]
For adding users on image repository creation current implementation is fine, but in the nearest future when we need to support changing users list dynamically for all image repositories, this approach is not optimal. Better way would be to create a team per workspace with the configmap and add users dynamically to the team.
But for this PR you don't have to rework it.

File: controllers/imagerepository_controller.go [hunk 1]
```diff
@@ -343,6 +346,11 @@ func (r *ImageRepositoryReconciler) ProvisionImageRepository(ctx context.Context
 		}
 	}
 
+	err = r.GrantAdditionalRepositoryAccess(ctx, imageRepository)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/imagerepository_controller.go:350]: ok

File: controllers/imagerepository_controller.go [hunk 2]
```diff
@@ -429,6 +437,51 @@ func (r *ImageRepositoryReconciler) ProvisionImageRepositoryAccess(ctx context.C
 	return data, nil
 }
 
+func (r *ImageRepositoryReconciler) GrantAdditionalRepositoryAccess(ctx context.Context, imageRepository *imagerepositoryv1alpha1.ImageRepository) error {
+	log := ctrllog.FromContext(ctx).WithName("GrantAdditionalRepositoryAccess")
+	ctx = ctrllog.IntoContext(ctx, log)
```
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/imagerepository_controller.go:442]: removing, not actually even sure what it does

File: controllers/imagerepository_controller.go [hunk 3]
```diff
@@ -429,6 +437,51 @@ func (r *ImageRepositoryReconciler) ProvisionImageRepositoryAccess(ctx context.C
...
+				log.Info("failed to add permissions for account, because it doesn't exist", "AccountName", user)
+				continue
+			}
+
+			log.Error(err, "failed to add permissions for account", "AccountName", user, l.Action, l.ActionUpdate, l.Audit, "true")
+			return err
+		}
+		log.Info("Additional user access was granted for", "UserName", user)
+	}
+	listPerms, _ := r.QuayClient.ListPermissionsForRepository(r.QuayOrganization, imageRepositoryName)
```
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/imagerepository_controller.go:477]: I'll remove the call, it was mostly for debugging

and I'll need that method for my other PR

[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/imagerepository_controller.go:477]: will remove the call

File: controllers/imagerepository_controller.go [hunk 4]
```diff
@@ -429,6 +437,51 @@ func (r *ImageRepositoryReconciler) ProvisionImageRepositoryAccess(ctx context.C
...
+	log.Info("Repository has now following permissions", "RepositoryName", imageRepositoryName)
```
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/imagerepository_controller.go:478]: removed with method call

File: controllers/imagerepository_controller.go [hunk 5]
```diff
@@ -429,6 +437,51 @@ func (r *ImageRepositoryReconciler) ProvisionImageRepositoryAccess(ctx context.C
...
+	for _, user := range listPerms {
+		log.Info(fmt.Sprintf("name: %s; role: %s; is robot: %t; is org member: %t", user.Name, user.Role, user.IsRobot, user.IsOrgMember))
```
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/imagerepository_controller.go:480]: removed

File: pkg/quay/quay.go [hunk 6]
```diff
@@ -359,18 +360,46 @@ func (c *QuayClient) DeleteRobotAccount(organization string, robotName string) (
...
+	}
+	var response Response
+	if err := resp.GetJson(&response); err != nil {
+		return nil, err
+	}
+
+	return response.Permissions, nil
+}
+
+// AddPermissionsForRepositoryToAccount allows given robot account to access to the given repository.
```
[the remark this answers is not in the document]
Comment (rcerven) [reply pkg/quay/quay.go:387]: changed

File: controllers/suite_util_test.go [hunk 7]
```diff
@@ -348,3 +349,30 @@ func deleteServiceAccount(serviceAccountKey types.NamespacedName) {
 		return k8sErrors.IsNotFound(k8sClient.Get(ctx, serviceAccountKey, serviceAccount))
 	}, timeout, interval).Should(BeTrue())
 }
+
+func createUsersConfigMap(configMapKey types.NamespacedName, users []string) {
+	configMapData := map[string]string{}
+	configMapData[additionalUsersConfigMapKey] = strings.Join(users, " ")
```
[the remark this answers is not in the document]
Comment (rcerven) [reply controllers/suite_util_test.go:355]: strings.Fields   splits string on ANY space / \n / \t

I've tried that during my testing

**Review by mmorhun — APPROVED** [review]

**Pull request merged**